### ISSUE TITLE
Allow passing exposed stan functions to the inner most function to avoid global environment issue.

### ISSUE THREAD

**Issue by Sandhu-SS** [paul-buerkner/brms#1577]
Hi, 

While working with the exposed stan function, it seems things work only if the exposed objects are in the global environment (default behavior in ``` brms::expose_functions()``` where ``` env = globalenv()```). 

Is it possible to add an argument to functions like ```fitted()``` and ```predict()``` to pass the exposed function all the way down to the inner most ```predictor()```? The name of this argument could **fun** (just an example, could be any other better name) which is NULL by default.  

This is to avoid the global environment issues raised by CRAN when using ```fitted()``` and ```predict()``` functions in R package examples.

Below is the rough outline of sequence of function called by the ```fitted.brmsfit()```

```fitted()``` -> ```prepare_predictions()``` -> ```posterior_epred()``` -> ```get_dpar()``` -> ```predictor()``` -> ```predictor.bprepnl()``` (```predictor.bprepnl()``` in my case as model if a nonlinear model). 

The ```predictor.bprepnl()```  is the function where exposed function (**prep$nlform**) is actually evaluated. Relevant code lines are:

```
 .fun <- function(...) eval(prep$nlform, list(...))
    eta <- try(
      t(do_call(mapply, c(list(FUN = .fun, SIMPLIFY = "array"), args))),
      silent = TRUE
    )
  } else {
    # assumes fully vectorized version of 'nlform'
    eta <- try(eval(prep$nlform, args), silent = TRUE)
  }
```

Thanks

**Comment by paul-buerkner** [comment]
how about using env = as.namespace("brms"). then the function should be
assigned to the brms namespace

Sandhu-SS ***@***.***> schrieb am Do., 21. Dez. 2023, 12:30:

**Comment by Sandhu-SS** [comment]
Thank you for your quick response,

I think the environment idea you suggested ```asNamespace("brms")``` is the best approach but unfortunately it does not work and throws the following error:

```
env <- asNamespace("brms")
brms::expose_functions(model, Vectorize = T, env = env  )

Error in rstan::expose_stan_functions(stanmodel, env = env, ...) : 
  Compilation failed!

```

**Comment by paul-buerkner** [comment]
I need a minimal reproducible example for your problem.

Sandhu-SS ***@***.***> schrieb am Do., 21. Dez. 2023, 15:04:

**Comment by Sandhu-SS** [comment]
Please see below an example:

```
library(brms)

# Mimic the package function by calling the nested functions, i.e., function_2 within the function_1 
# This way, setting the parent.frame() environment for the brms::expose_functions() is not .GlobalEnv

function_1 <- function() {
  
  function_2 <- function() {
    b <- c(2, 0.75)
    x <- rnorm(100)
    y <- rnorm(100, mean = b[1] * exp(b[2] * x))
    data <- data.frame(x, y)
    
    # Define Stan function 'customfun'    
    fun_code <- "
      vector customfun(vector b1, vector b2, vector x) {
         return (b1 .* exp(b2 .* x));
      }
    "    
    stanvars <- stanvar(scode = fun_code, block = "functions")
    
    prior <- prior(normal(1, 2), nlpar = "b1") + prior(normal(0, 2), nlpar = "b2")
    
    model <- brm(bf(y ~ customfun(b1, b2, x), 
                    b1 + b2 ~ 1, nl = TRUE, loop = FALSE),
                 stanvars = stanvars,
                 prior = prior,
                 data = data, 
                 chains = 1)
    
    # env = parent.frame() does not work ->  This because 'customfun' not available in the .GlobalEnv
     expose_functions(model, env = parent.frame())
     
    # env = parent.frame() does not work -> Compilation failed!
     expose_functions(model, env = asNamespace('brms'))
     
    # Default env = globalenv() works but 'customfun' is exposed to global env
     expose_functions(model, env =  globalenv())
     
    out <- fitted(model)
    return(out)
  }
  out2 <- function_2()
  return(out2)
}

# Call function_1 to fit model and get fitted values via fitted()
# It returns fitted values only when env = globalenv()  in function_2

function_1()
```


Thank you

**Comment by paul-buerkner** [comment]
The error message was misleading. After I fixed the workflow of expose_function with the rstan backend, I got the correct informative error which says "cannot add bindings to a locked environment". So basically, the brms environment was locked upon loading the package. Apparently rlang can unlock environments via `env_binding_unlock` but it didn't work for me. 

Anyway, I now found a way to convince brms to look for functions in the environment of the model formula, that is where the model was originally run. This is consistent with usually R formula behavior and fixes your cases if you run `expose_functions(model, env = environment())` that is expose the function in the current environment
